src/coreclr/vm/appdomain.cpp

```diff
@@ -3778,39 +3769,41 @@ void AppDomain::AddUnmanagedImageToCache(LPCWSTR libraryName, NATIVE_LIBRARY_HAN
     CONTRACTL
     {
         THROWS;
-        GC_TRIGGERS;
+        GC_NOTRIGGER;
         MODE_ANY;
         PRECONDITION(CheckPointer(libraryName));
+        PRECONDITION(CheckPointer(hMod));
         INJECT_FAULT(COMPlusThrowOM(););
     }
     CONTRACTL_END;
-    if (libraryName)
-    {
-        AssemblySpec spec;
-        spec.SetCodeBase(libraryName);
-        m_UnmanagedCache.InsertEntry(&spec, hMod);
-    }
-    return ;
-}
 
+    DWORD hash = HashString(libraryName);
```

Comment:
We are using the hash code as the key in the MapSHash. I do not think that works. The hashcode can have collisions.

Reply:
I made it only add if it doesn't already exist and assert that the existing handle is the same as the one passed in if it does exist.

I believe this should be the same as the existing behaviour, since we were just using an `AssemblySpec` with the code base set, so the hash of the assembly spec was just the hash string of the code base.

Reply:
What happens when we end up with two different library names that HashString returns the same hash code for?

>  the hash of the assembly spec was just the hash string of the code base.

The assemblyspec table uses the whole assemblyspec as the key. It does not use just the assemblyspec hashcode as the key.

Reply:
Isn't the `DomainAssemblyCache` just passing in the assembly spec hash (or the AssemblyEntry hash which returns the assembly spec hash) as the key on lookup and insert?

https://github.com/dotnet/runtime/blob/f39d544d6dcbeaf2093657336b06f8712e9b3f6b/src/coreclr/vm/assemblyspec.cpp#L1600-L1602

https://github.com/dotnet/runtime/blob/f39d544d6dcbeaf2093657336b06f8712e9b3f6b/src/coreclr/vm/assemblyspec.cpp#L1643-L1644

Or is that not what you are talking about?

Reply:
`PtrHashMap` used to implement `DomainAssemblyCache` has super weird design. The argument named `key` is actually the hashcode, and the argument named `value` is actually the key in the `DomainAssemblyCache` case.

Reply:
Ergh, I had stepped through how the key (argument) was set/used without fully going through everything for the value (argument). I see it now - thanks.